src/backtracking/permutations.rs

```diff
@@ -1,63 +1,137 @@
-/*
-The permutations problem involves finding all possible permutations
-of a given collection of distinct integers. For instance, given [1, 2, 3],
-the goal is to generate permutations like
- [1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], and [3, 2, 1].
- This implementation uses a backtracking algorithm to generate all possible permutations.
-*/
-
-pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
-    let mut result = Vec::new(); // Vector to store the resulting permutations
-    let mut current_permutation = Vec::new(); // Vector to store the current permutation being constructed
-    let mut used = vec![false; nums.len()]; // A boolean array to keep track of used elements
-
-    backtrack(&nums, &mut current_permutation, &mut used, &mut result); // Call the backtracking function
-
-    result // Return the list of permutations
+//! This module provides a function to generate all possible distinct permutations
+//! of a given collection of integers using a backtracking algorithm.
+
+/// Generates all possible distinct permutations of a given vector of integers.
+///
+/// # Arguments
+///
+/// * `nums` - A vector of integers.
+///
+/// # Returns
+///
+/// A vector containing all possible distinct permutations of the input vector.
+pub fn permute(mut nums: Vec<isize>) -> Vec<Vec<isize>> {
+    let mut permutations = Vec::new();
+    let mut current = Vec::new();
+    let mut used = vec![false; nums.len()];
+
+    nums.sort();
+    generate(&nums, &mut current, &mut used, &mut permutations);
+
+    permutations
 }
 
-fn backtrack(
-    nums: &Vec<i32>,
-    current_permutation: &mut Vec<i32>,
+/// Helper function for the `permute` function to generate distinct permutations recursively.
+///
+/// # Arguments
+///
+/// * `nums` - A reference to the original vector of integers.
+/// * `current` - A mutable reference to the vector holding the current permutation.
+/// * `used` - A mutable reference to a vector tracking which elements are used.
+/// * `permutations` - A mutable reference to the vector holding all generated distinct permutations.
+fn generate(
+    nums: &Vec<isize>,
```

Comment:
You could use a slice here:

```suggestion
    nums: &[size],
```

This will also require an update of the doc-string.

Reply:
Thanks.